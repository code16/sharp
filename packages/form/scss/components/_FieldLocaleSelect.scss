.SharpFieldLocaleSelect {
  margin-bottom: 0.625rem;
  &__btn {
    text-transform: uppercase;
    background: $ui-01;
    border: 1px solid $ui-03;
    color: $text-02;
    border-radius: 5px;
    -webkit-appearance: none;
    outline: 0;
    cursor: pointer;

    &:hover {
      border: 1px solid $brand-01;
    }
  }
  &__btn--empty {
    border: 1px dashed $brand-01;
    background: transparent;
  }
  &__btn--active {
    background: $brand-01;
    color: $inverse-01;
  }
}