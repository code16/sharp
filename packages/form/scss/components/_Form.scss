@import "../mixins/fix-module-fields-margin";

.SharpForm {

  .close {
    font-weight: normal;
  }

  &__fieldset {
    @include reset;
    //margin-bottom: 2rem;

    @extend .SharpModule;
    @include fix-module-fields-margin;

    &>.SharpModule__inner {
      background: transparent;
      border: 0;
      @include layer('base');
      &>.SharpModule__header {
        padding: 0;
        .SharpModule__title {
          @extend .SharpForm__label;
        }
      }
      &>.SharpModule__content {
        background: $ui-03;
        @include layer('raised');
      }
    }
  }

  .SharpGrid__col--empty {
    display: none;
  }

  .SharpGrid__col:not(.SharpGrid__col--empty), &__form-item--row {
    margin-bottom: $form-item-margin-bottom;
  }

  &__form-item {
    flex: 1;
    height: 100%;
    //margin-right: 1rem;

    /*&:last-of-type {
      margin-right: 0;
      margin-bottom: 0;
    }*/

    // For single check align with other field
    &--type-check {
      .SharpForm__label {
        &::after {
          content: "\a0\a0"
        }
      }
      .SharpCheck {
        height: 2.5rem;
      }
    }
  }

  &__label {
    @include reset;
    @include font-size('14');
    font-weight: $input-label-weight;
    display: inline-block;
    vertical-align: baseline;
    margin-bottom: rem(10px) !important;
    text-transform: uppercase;
  }

  &__label--disabled {
    opacity: .5;
  }

  &__form-item--danger {
    input, textarea, select, .SharpDropdown__text {
      //border-bottom: 2px solid $support-01;
    }
    .SharpForm__form-requirement {
      max-height: rem(200px);
      background: rgba($support-01, .1);
      margin: 0;
      padding: .5rem 1rem;
    }
  }


  &__form-requirement {
    @include reset;
    @include font-size('11');
    margin: .75rem 0 0;
    max-height: 0;
    overflow: hidden;
    font-weight: 700;
    line-height: 1.5;

    &:empty {
      display:none;
    }

    &::before {
      content: '*';
      display: inline-block;
      color: $support-01;
    }
  }

  &__help-message {
    @include font-size('11');
    margin: .75rem 0 0;
    line-height: 1.5;
  }



  &__locale-dropdown {
    text-transform: uppercase;
    width: 4rem;
  }

  .SharpTabs {
    .SharpForm__locale-dropdown {
      align-self: center;
      margin-right: 1rem;
      @include media-breakpoint-up(md) {
        position: absolute;
        left: 0;
        top: 0;
        margin-right: 0;
      }
    }
    &--nav-overflow {
      .SharpForm__locale-dropdown {
        order: 1;
        margin-top: .5rem;
        align-self: flex-start;
      }
      .SharpTabs__collapse-btn {
        align-self: flex-start;
      }
    }
  }

  /// reset multiselect new Design
  .multiselect__tags {
    border-radius: 0;
    border: none;
  }

  .multiselect__option:not(:hover):not(.multiselect__option--selected):not(.multiselect__option--highlighted)
  {
      background: white!important;
  }
}