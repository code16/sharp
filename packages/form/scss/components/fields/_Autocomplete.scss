.SharpAutocomplete {
  position: relative;
  font-size: 1rem;

  &--disabled {
    opacity: .5;
    cursor: not-allowed !important;
    outline: none !important;
    .SharpAutocomplete__result-item__close-button {
      pointer-events: none;
      cursor: not-allowed !important;
    }
  }

  &__result-item {
    &__close-button {
      @extend %select-close-button;
    }

    &__close-icon {
      @extend %select-close-icon;
    }
  }

  &__multiselect {
    &--hide-dropdown.multiselect--active {
      .multiselect__content {
        display: none !important;
      }
    }
    .multiselect__tags {
      padding-right: 2rem !important;
    }
    &.multiselect--active {
      .multiselect__input {
        width: 100%!important;
      }
    }
    .multiselect__option > .SharpTemplate {
      overflow:hidden;
      text-overflow: ellipsis;
    }
    .multiselect__single {
      .SharpTemplate {
        overflow:hidden;
        text-overflow: ellipsis;
      }
    }
  }

  &__overlay {
    position: absolute;
    left: 0;
    top: 0;
    right: 2rem;
    bottom: 0;
    width: auto;

    .multiselect__tags {
      padding-right: 0;
      box-shadow: none!important;
      user-select: auto!important;
    }
  }

  &--remote {
    .multiselect__select {
      display: none;
    }

    .multiselect__tags {
      padding-right: 1rem;
    }
  }

  .SharpLoading {
    position: absolute;
    right: 0;
    top: 50%;
    transform: translateY(-50%);
    &__svg {
      stroke: $brand-01
    }
  }
}