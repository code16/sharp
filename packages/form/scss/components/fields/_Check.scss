.SharpCheck {
  height: 100%;
  &__label {
    @include reset;
    @include font-smoothing;
    @include font-size('14');
    display: flex;
    height: 100%;
    align-items: center;
    cursor: pointer;

    .SharpCheck__input {
      @include hidden;
    }

    .SharpCheck__label-text {
      user-select: none;
    }

    .SharpCheck__appearance {
      display: inline-block;
      height: rem(18px);
      min-width: rem(18px);
      max-width: rem(18px);
      margin-right: .5rem;
      background-color: $ui-01;
      border-radius: 2px;
      border: 2px solid $ui-05;
    }

    .SharpCheck__input:checked + .SharpCheck__appearance,
    .SharpCheck__input:checked + .SharpCheck__label .SharpCheck__appearance {
      display: flex;
      align-items: baseline;
      background-color: $brand-01;
      border-color: $brand-01;
    }

    .SharpCheck__input:disabled + .SharpCheck__appearance,
    .SharpCheck__input:disabled ~ .SharpCheck__label-text {
      opacity: .5;
      cursor: not-allowed;
    }

    .SharpCheck__input:focus + .SharpCheck__appearance {
      @include focus-outline('blurred');
    }

    .SharpCheck__checkmark {
      display: none;
      fill: $inverse-01;
      width: rem(15px);
      height: rem(15px);
    }

    .SharpCheck__input:checked + .SharpCheck__appearance .SharpCheck__checkmark {
      display: block;

      @-moz-document url-prefix() {
        & {
          stroke: $brand-01;
        }
      }
    }
  }
}