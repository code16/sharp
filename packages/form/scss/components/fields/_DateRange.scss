@import '~element-ui/packages/theme-chalk/src/date-picker';

.SharpDateRange {
  .el-range__close-icon {
    display: none;
    font-weight: normal;
    font-size: 16px;
    -webkit-font-smoothing: subpixel-antialiased;
    font-family: inherit!important;
    &::before {
      content: 'Ã—';
    }
  }
  &--clearable {
    .el-range__close-icon {
      display: inline-block;
    }
  }
}