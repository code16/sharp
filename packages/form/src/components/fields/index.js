import Autocomplete from './Autocomplete.vue';
import Textarea from './Textarea.vue';
import Text from './Text.vue';
import Markdown from './markdown/Markdown.vue';
import NumberInput from './Number.vue';
import Upload from './upload/Upload.vue';
import TagInput from './Tags.vue';
import DateInput from './date/Date.vue';
import Check from './Check.vue';
import List from './list/List.vue';
import Select from './Select.vue';
import Html from './Html.vue';
import Geolocation from './geolocation/Geolocation';
import Trix from './wysiwyg/TrixEditor.vue';
import DateRange from './date-range/DateRange';

export default {
    'autocomplete' : Autocomplete,
    'text'         : Text,
    'textarea'     : Textarea,
    'markdown'     : Markdown,
    'number'       : NumberInput,
    'upload'       : Upload,
    'tags'         : TagInput,
    'date'         : DateInput,
    'check'        : Check,
    'list'         : List,
    'select'       : Select,
    'html'         : Html,
    'geolocation'  : Geolocation,
    'wysiwyg'      : Trix,
    'daterange'    : DateRange,
};

export {
    Autocomplete,
    Text,
    Textarea,
    Markdown,
    NumberInput,
    Upload,
    TagInput,
    DateInput,
    Check,
    List,
    Select,
    Html,
    Geolocation,
    Trix,
    DateRange,
}