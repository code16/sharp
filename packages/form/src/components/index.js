
export * from './fields';
export { default as Form } from './Form';