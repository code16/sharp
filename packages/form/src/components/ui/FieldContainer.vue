<template>
    <div class="SharpFieldContainer SharpForm__form-item" :class="formGroupClasses" :style="extraStyle">
        <div class="row">
            <div class="col">
                <label v-if="showLabel" class="SharpForm__label" @click="triggerFocus">
                    {{label}}
                </label>
            </div>
            <template v-if="fieldProps.localized">
                <div class="col-auto">
                    <FieldLocaleSelect
                        :locales="$form.locales"
                        :current-locale="locale"
                        :field-value="resolvedOriginalValue"
                        :is-locale-object="isLocaleObject"
                        @change="handleLocaleChanged"
                    />
                </div>
            </template>
        </div>
        <Field v-bind="exposedProps"
            @error="setError"
            @ok="setOk"
            @clear="clear"
            @blur="handleBlur"
            ref="field"
        />
        <div class="SharpForm__form-requirement">{{stateMessage}}</div>
        <small class="SharpForm__help-message">{{helpMessage}}</small>
    </div>
</template>

<script>
    import { logError } from 'sharp';
    import { ErrorNode, ConfigNode }  from 'sharp/mixins';
    import Field from '../Field';
    import FieldLocaleSelect from './FieldLocaleSelect';
    import { resolveTextValue, isLocalizableValueField } from '../../util/locale';


    export default {
        name: 'SharpFieldContainer',

        mixins: [ ErrorNode, ConfigNode ],

        components: {
            Field,
            FieldLocaleSelect,
        },

        inject:['$tab', '$form'],

        props : {
            ...Field.props,

            label: String,
            helpMessage: String,
            originalValue: [String, Number, Boolean, Object, Array, Date],
        },
        data() {
            return {
                state:'classic',
                stateMessage:''
            }
        },
        watch: {
            value() {
                if(this.state === 'error')
                    this.clear();
            },
            '$form.errors'(errors) {
                this.updateError(errors);
            },
            locale() {
                this.updateError(this.$form.errors);
            }
        },
        computed: {
            formGroupClasses() {
                return [
                    `SharpForm__form-item--type-${this.fieldType}`,
                    {
                        'SharpForm__form-item--danger': this.state==='error',
                        'SharpForm__form-item--success': this.state==='ok',
                        'SharpForm__form-item--no-label': !this.showLabel,
                    }
                ];
            },
            extraStyle() {
                return this.fieldProps.extraStyle;
            },
            exposedProps() {
                return {
                    ...this.$props,
                    uniqueIdentifier: this.mergedErrorIdentifier,
                    fieldConfigIdentifier: this.mergedConfigIdentifier
                };
            },
            showLabel() {
                return !!this.label || this.label === '';
            },
            resolvedOriginalValue() {
                return resolveTextValue({ field:this.fieldProps, value:this.originalValue });
            },
            isLocaleObject() {
                return isLocalizableValueField(this.fieldProps);
            }
        },
        methods: {
            updateError(errors) {
                let error = errors[this.mergedErrorIdentifier];
                if(error == null) {
                    this.clear();
                }
                else if(Array.isArray(error)) {
                    this.setError(error[0]);
                }
                else {
                    logError(`FieldContainer : Not processable error "${this.mergedErrorIdentifier}" : `, error);
                }
            },
            setError(error) {
                this.state = 'error';
                this.stateMessage = error;
                if(this.$tab) {
                    this.$tab.$emit('error', this.mergedErrorIdentifier);
                }
            },
            setOk() {
                this.state = 'ok';
                this.stateMessage = '';
            },
            clear() {
                this.state = 'classic';
                this.stateMessage = '';
                if(this.$tab) {
                    this.$tab.$emit('clear', this.mergedErrorIdentifier);
                }
                this.$form.$emit('error-cleared', this.mergedErrorIdentifier);
            },
            triggerFocus() {
                this.$set(this.fieldProps,'focused',true);
            },
            handleBlur() {
                this.$set(this.fieldProps,'focused',false);
            },
            handleLocaleChanged(locale) {
                this.$emit('locale-change', this.fieldKey, locale);
            }
        },
        mounted() {
            //console.log(this);
        }
    }
</script>