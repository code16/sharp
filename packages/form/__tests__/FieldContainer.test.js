import Vue from 'vue';
import { ErrorNode, ConfigNode } from 'sharp/mixins';
import FieldContainer from '../src/components/ui/FieldContainer.vue';
import * as fields from '../src/components/fields';

import { MockInjections } from "@sharp/test-utils";


describe('field-container', () => {
    Vue.component('sharp-field-container', FieldContainer);

    beforeEach(() => {
        document.body.innerHTML = `
            <div id="app">
                <sharp-field-container
                    field-key="title"
                    field-type="test-field"
                    :field-props="fieldProps || {}"
                    :locale="locale"
                    label="My beautiful field"
                    help-message="Do not fill this field"
                    error-identifier="title"
                    config-identifier="title"
                >
                </sharp-field-container> 
            </div>
        `;

        fields.default['test-field'] = {
            template: '<div>Template field</div>'
        };
    });

    test('can mount field container', async () => {
        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount field container with error', async () => {
        let $fieldContainer = await createVm();

        $fieldContainer.$form.errors = {
            'error.title': [
                'Must be a number'
            ]
        };
        await Vue.nextTick();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount field container with success', async () => {
        let $fieldContainer = await createVm();

        $fieldContainer.state = 'ok';

        await Vue.nextTick();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount field container with extra style', async () => {
        await createVm({
           data: ()=>({
                fieldProps: {
                    extraStyle: 'background:red'
                }
            })
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('set error on field emitted error', async () => {
        let $fieldContainer = await createVm();

        let { field } = $fieldContainer.$refs;
        let tabErrorHandler = jest.fn();

        $fieldContainer.$tab.$on('error', tabErrorHandler);

        field.$emit('error', 'Error message');

        expect($fieldContainer.state).toBe('error');
        expect($fieldContainer.stateMessage).toBe('Error message');
        expect(tabErrorHandler).toHaveBeenCalledWith('error.title');
    });

    test('set success on field emitted ok', async () => {
        let $fieldContainer = await createVm();

        let { field } = $fieldContainer.$refs;

        field.$emit('ok');
        expect($fieldContainer.state).toBe('ok');
        expect($fieldContainer.stateMessage).toBe('');
    });

    test('clear on clear emitted', async () => {
        let $fieldContainer = await createVm();

        let { field } = $fieldContainer.$refs;
        let tabClearHandler = jest.fn(), formClearHandler = jest.fn();

        $fieldContainer.$tab.$on('clear', tabClearHandler);
        $fieldContainer.$form.$on('error-cleared', formClearHandler);
        field.$emit('clear');

        expect($fieldContainer.state).toBe('classic');
        expect($fieldContainer.stateMessage).toBe('');
        expect(tabClearHandler).toHaveBeenCalledWith('error.title');
        expect(formClearHandler).toHaveBeenCalledWith('error.title');
    });

    test('expose proper props', async () => {
        let $fieldContainer = await createVm();

        let { field } = $fieldContainer.$refs;
        expect($fieldContainer.exposedProps).toMatchObject(field.$props);
        expect(field.uniqueIdentifier).toBe('error.title');
        expect(field.fieldConfigIdentifier).toBe('config.title');
    });

    test('responsive to $form.errors object', async () => {
        let $fieldContainer = await createVm();

        $fieldContainer.clear = jest.fn();
        $fieldContainer.setError = jest.fn();

        let { $form } = $fieldContainer;

        $form.errors = {
            'error.title': null
        };
        await Vue.nextTick();
        expect($fieldContainer.clear).toHaveBeenCalled();

        $form.errors = {
            'error.title': ['Error message']
        };
        await Vue.nextTick();
        expect($fieldContainer.setError).toHaveBeenCalledWith('Error message');
    });

    test('update errors on locale changed', async () => {
        let $fieldContainer = await createVm({
            data:()=>({locale:'en'})
        });
        let { $root:vm, $form } = $fieldContainer;

        $fieldContainer.updateError = jest.fn();
        $form.errors = { error1: 'text' };
        vm.locale = 'fr';

        await Vue.nextTick();
        expect($fieldContainer.updateError).toHaveBeenCalledWith({ error1: 'text' });
    });
});

async function createVm(customOptions={}) {

    const vm = new Vue({
        el: '#app',
        mixins: [MockInjections, customOptions, ErrorNode, ConfigNode],

        'extends': {
            data:()=>({
                fieldProps: {},
                locale: null
            }),
            propsData: {
                errorIdentifier: 'error',
                configIdentifier: 'config'
            },
        }
    });

    await Vue.nextTick();

    return vm.$children[0];
}