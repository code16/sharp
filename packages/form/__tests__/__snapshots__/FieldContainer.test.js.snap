// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field-container can mount field container 1`] = `
<div id="app">
	<div class="SharpFieldContainer SharpForm__form-item SharpForm__form-item--type-test-field">
		<div class="row">
			<div class="col">
				<label class="SharpForm__label">
					My beautiful field
				</label>
			</div>
			<!---->
		</div>
		<div>Template field</div>
		<div class="SharpForm__form-requirement"></div>
		<small class="SharpForm__help-message">Do not fill this field</small>
	</div>
</div>
`;

exports[`field-container can mount field container with error 1`] = `
<div id="app">
	<div class="SharpFieldContainer SharpForm__form-item SharpForm__form-item--type-test-field SharpForm__form-item--danger">
		<div class="row">
			<div class="col">
				<label class="SharpForm__label">
					My beautiful field
				</label>
			</div>
			<!---->
		</div>
		<div>Template field</div>
		<div class="SharpForm__form-requirement">Must be a number</div>
		<small class="SharpForm__help-message">Do not fill this field</small>
	</div>
</div>
`;

exports[`field-container can mount field container with extra style 1`] = `
<div id="app">
	<div class="SharpFieldContainer SharpForm__form-item SharpForm__form-item--type-test-field" style="background: red;">
		<div class="row">
			<div class="col">
				<label class="SharpForm__label">
					My beautiful field
				</label>
			</div>
			<!---->
		</div>
		<div>Template field</div>
		<div class="SharpForm__form-requirement"></div>
		<small class="SharpForm__help-message">Do not fill this field</small>
	</div>
</div>
`;

exports[`field-container can mount field container with success 1`] = `
<div id="app">
	<div class="SharpFieldContainer SharpForm__form-item SharpForm__form-item--type-test-field SharpForm__form-item--success">
		<div class="row">
			<div class="col">
				<label class="SharpForm__label">
					My beautiful field
				</label>
			</div>
			<!---->
		</div>
		<div>Template field</div>
		<div class="SharpForm__form-requirement"></div>
		<small class="SharpForm__help-message">Do not fill this field</small>
	</div>
</div>
`;
